src/status_im/chat/db.cljs

```diff
@@ -64,27 +73,86 @@
 
+(defn add-datemark
+  [[datemark message-references]]
+  (conj message-references
+        {:value datemark
+         :type  :datemark}))
+
+(defn datemark? [{:keys [type]}]
+  (= type :datemark))
+
+(defn transform-message
+  [messages message-statuses referenced-messages]
+  (fn [{:keys [message-id timestamp-str] :as reference}]
+    (if (datemark? reference)
+      reference
+      (let [{:keys [content] :as message} (get messages message-id)
+            {:keys [response-to response-to-v2]} content
+            quote (some-> (or response-to-v2 response-to)
+                          (quoted-message-data messages referenced-messages))]
+        (cond-> (-> message
+                    (update :content dissoc :response-to :response-to-v2)
+                    (assoc :timestamp-str timestamp-str
+                           :user-statuses (get message-statuses message-id)))
+          ;; quoted message reference
+          quote
+          (assoc-in [:content :response-to] quote))))))
+
+(defn check-gap
+  [{:keys [exists? from]} previous-message next-message gap-added?]
+  (let [previous-timestamp (:whisper-timestamp previous-message)
+        next-timestamp     (:whisper-timestamp next-message)]
+    (and (not gap-added?)
+         (or (and exists? previous-timestamp next-timestamp
+                  (< previous-timestamp from next-timestamp))
+             (and exists? (nil? next-message))))))
+
+(defn add-gap [messages gap]
+  (conj messages
+        {:type  :gap
+         :value (str (:from gap))}))
+
 (defn messages-with-datemarks-and-statuses
   "Converts message groups into sequence of messages interspersed with datemarks,
   with correct user statuses associated into message"
-  [message-groups messages message-statuses referenced-messages]
-  (mapcat (fn [[datemark message-references]]
-            (into (list {:value datemark
-                         :type  :datemark})
-                  (map (fn [{:keys [message-id timestamp-str]}]
-                         (let [{:keys [content] :as message} (get messages message-id)
-                               {:keys [response-to response-to-v2]} content
-                               quote (some-> (or response-to-v2 response-to)
-                                             (quoted-message-data messages referenced-messages))]
-                           (cond-> (-> message
-                                       (update :content dissoc :response-to :response-to-v2)
-                                       (assoc :datemark      datemark
-                                              :timestamp-str timestamp-str
-                                              :user-statuses (get message-statuses message-id)))
-                             quote ;; quoted message reference
-                             (assoc-in [:content :response-to] quote)))))
-                  message-references))
-          message-groups))
+  [message-groups messages message-statuses referenced-messages messages-gap]
```

Comment:
I believe we have some tests already for this, they should probably be updated to reflect the changes and test the new functionalities

Reply:
ok

Reply:
added